### PR TITLE
[CA-211267] Add a lightweight polling mechanism to the appliance tasks, to provide an accurate timer

### PR DESCRIPTION
**Pull request by mcintyre94**
For other actions we get a ticking timer in the Events view because they use PollToCompletion, which calls their Changed event (via Action.Tick) every 900ms. These actions don't use PollToCompletion, and didn't regularly call their Changed event - so their timer didn't tick regularly. This change adds a lightweight mechanism to call their changed event every 900ms until they complete. This allows the events page timer to tick every second in the same way as it does for other actions.

Signed-off-by: Callum McIntyre callumiandavid.mcintyre@citrix.com

### CODE REVIEW

**Comment by Frezzle** [comment]
should this be CA-211267?

**Comment by mcintyre94** [comment]
Okay, renamed the first commit and moved the sleep timer to const int in the class.

File: XenAdmin/Actions/OVFActions/ApplianceAction.cs [hunk 1]
```diff
@@ -97,5 +100,21 @@ protected override void CancelRelatedTask()
...
 		}
+
+	    protected void InitialiseTicker()
+	    {
+	        System.Threading.Tasks.Task.Run(() => TickUntilCompletion());
+	    }
+
+	    private void TickUntilCompletion()
+	    {
+	        while (!IsCompleted)
```
Comment (kc284) [inline XenAdmin/Actions/OVFActions/ApplianceAction.cs]: Could we add a null check here? I'm a bit paranoid the action might be nullified on a different thread and this one can't notice.

File: XenAdmin/Actions/OVFActions/ExportApplianceAction.cs [hunk 2]
```diff
@@ -92,6 +92,9 @@ protected override void Run()
 		{
 		    SafeToExit = false;
 			var session = Connection.Session;
+
+            InitialiseTicker();
```
Comment (kc284) [inline XenAdmin/Actions/OVFActions/ExportApplianceAction.cs]: Personally I'd rather the child classes didn't have to call the ticker, e.g. I would write in the parent class:
```
protected override void Run()
{
    SafeToExit = false;
    InitialiseTicker();
}
```
and in the child class
```
protected override void Run()
{
    base.Run();
    everythingelse();
}
```

**Comment by stephen-turner** [comment]
Sorry, I'm confused why that last change is needed. "this" can never be null, can it?

**Comment by kc284** [comment]
Apologies for the delay, I had to find the relevant article http://stackoverflow.com/questions/10625326/this-null-inside-net-instance-method-why-is-that-possible

**Comment by mcintyre94** [comment]
I think something like:

var action = new ExportApplianceAction(...);
action.Run(); // starts timer thread
action = null;

Would cause a null dereference. I don't think there's anything we can do to stop the creator of an action suddenly making it null, and I don't think the spawned thread would be destroyed by that, so it'd keep trying to use the null action.

**Comment by stephen-turner** [comment]
No, I don't think that's the case. My reading of that StackOverflow thread and others is that "this" can only be null if you use reflection, or fiddle with the IL. An ordinary call to a member function cannot have "this" as null, and the function will continue running with the same "this" until it exits: another thread changing its "action" variable to null doesn't suddenly change "this" in the function that's running. Otherwise, you'd have to keep testing for (this != null) before any (explicit or implicit) dereferencing of "this", including any read of a data member.

If you're not convinced, your example is easy to test.

**Comment by GaborApatiNagy** [comment]
"action = null;" this will not make the instance null or will not remove it at all. It will only make 'action' to point to nowhere, the instance will still be in the memory until the garbage collector decides to clear it. I haven't tried, but I assume it won't, as the events are still subscribed in it.

**Comment by stephen-turner** [comment]
What I'm more worried about in this code is, is the action guaranteed to reach the "IsCompleted" state eventually? If not, the timer will keep ticking, the action cannot be GCed, and we have a memory leak.

**Comment by GaborApatiNagy** [comment]
I think if we had to start testing (this == null) everywhere, that would be a chaos :) and broken design somewhere...

**Comment by stephen-turner** [comment]
See Eric Lippert's answer in http://stackoverflow.com/questions/19300245/is-it-possible-for-this-keyword-to-equal-null:

"So there are several ways in which this could be null in C#, but they are all very much out of the mainstream. It is therefore very rare for people to write code as your professor has. C# programmers idiomatically suppose that this is not null and never check for it."

**Comment by GaborApatiNagy** [comment]
Just looking into this a bit more, I may just take a step back and ask why do we need to pass 'this' around at all?
Could we just have the following?

 private void InitialiseTicker()
+	    {
+	        System.Threading.Tasks.Task.Run(() => TickUntilCompletion());
+	    }
+
+	    private static  void TickUntilCompletion()
+	    {
+              #use this here
+         }

**Comment by mcintyre94** [comment]
That's how it was one commit ago (https://github.com/xenserver/xenadmin/pull/1236/commits/15b47059428f61e8c29a724e5363cfb3572d92f3), I'm happy to go back to that if the null issue isn't a concern.

**Comment by stephen-turner** [comment]
Yes, @GaborApatiNagy, that's exactly what the code was until the last changeset (7d7a040). I think the previous version was correct.

**Comment by stephen-turner** [comment]
@mcintyre94 Yes, you can revert 7d7a040, but I would like to be sure there isn't a memory leak before merging it.

**Comment by mcintyre94** [comment]
I didn't revert the whole commit, the rest of it is still worth having.

Some investigation on the reliabiltiy of Action.IsCompleted: I think it's okay if the events always run using the RunAsync() or RunExternal() methods. Those both use AsyncAction.RunWorkerThread(), which, by calling Action.MarkCompleted(), guarantees completion:

- At the end of Run()
- On the action throwing a CancelledException
- On any other caught Exception

The only case I can see where RunWorkerThread() won't mark an event as completed is where it has already been cancelled (RunWorkedThread just returns), but in this case the CancellingAction.Cancel() event will mark the action as completed instead. 

If the events are run using just their Run() method (which we don't currently for any of these events), then I don't think they will complete. Notably setting PercentComplete to 100 doesn't seem to trigger AsyncAction.IsCompleted, and they don't set IsCompleted themselves on exceptions.  I'm not sure if this is a concern though, because we don't and shouldn't run these synchronously.

**Comment by stephen-turner** [comment]
I'm wondering if we should have a 24-hour timeout, just to guard against some code path we don't understand or might accidentally add later.

Admittedly, a VM import can take more than 24 hours, but even if it does, it's not too bad a bug to stop the counter then.

**Comment by GaborApatiNagy** [comment]
This would do the same, but more meaningful

**Comment by GaborApatiNagy** [comment]
The reason I've suggested this is just to make sure 24h (approx, I know) stays 24h even if someone changes SLEEP_TIME_MS in the future.

**Comment by GaborApatiNagy** [comment]
:+1: Thanks, merging

**Pull request merged**